Enable omnibus build cache

chouquette opened the pull request:
<!--
* New contributors are highly encouraged to read our
  [CONTRIBUTING](/CONTRIBUTING.md) documentation.
* The pull request:
  * Should only fix one issue or add one feature at a time.
  * Must update the test suite for the relevant functionality.
  * Should pass all status checks before being reviewed or merged.
* Commit titles should be prefixed with general area of pull request's change.

-->
### What does this PR do?

This PR enables omnibus git cache, in order to stop rebuilding all of our dependencies during each CI job.
All packages before the agent are expected to rarely change, and shouldn't have to be rebuilt every single time. 
On average, this saves about 15/20 minutes per job.

This also allows individual developers to skip rebuilding every single dependencies if they wish to add a new software dependency to the agent. All they need to do is to provide the `OMNIBUS_GIT_CACHE_DIR` environment variable to a directory of their choosing.

<!--
* A brief description of the change being made with this pull request.
* If the description here cannot be expressed in a succinct form, consider
  opening multiple pull requests instead of a single one.
-->

### Motivation

This is part of the currently running initiative to reduce the median pipeline duration under 2 hours. This specific investigation is listed under https://datadoghq.atlassian.net/browse/APL-1805

Associated RFC: https://docs.google.com/document/d/1PSGpd2ixXXMbfzC1j0o514SXypcDiVamxBL3db__Bt0/edit?usp=sharing

<!--
* What inspired you to submit this pull request?
* Link any related GitHub issues or PRs here.
-->

### Additional Notes
<!--
* Anything else we should know when reviewing?
* Include benchmarking information here whenever possible.
* Include info about alternatives that were considered and why the proposed
  version was chosen.
-->

### Possible Drawbacks / Trade-offs

<!--
* What are the possible side-effects or negative impacts of the code change?
-->

### Describe how to test/QA your changes

<!--
* Write here in detail or link to detailed instructions on how this change can
  be tested/QAd/validated, including any environment setup.
-->

### Reviewer's Checklist
<!--
* Authors can use this list as a reference to ensure that there are no problems
  during the review but the signing off is to be done by the reviewer(s).

Note: Adding GitHub labels is only possible for contributors with write access.
-->

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature.
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

chouquette committed:
omnibus: conditionally enable git cache

chouquette committed:
tasks: omnibus_build: add support for omnibus git caching

chouquette committed:
omnibus: always build datadog projects/software

They are expected to almost always change and wouldn keep invalidating
the cache.
Not caching those will allow us to not regenerate the cache when there's
no need to, which saves a few minutes it takes to recreate the cache
bundle and upload it to s3

chouquette committed:
always update omnibus cache

so that we can measure the results until it's merged & further worked on

chouquette committed:
tasks: generate a cache key to fetch omnibus cache

chouquette committed:
currate env variables filter

chouquette committed:
attempt to stop hardcoding install dir

Pythyu committed:
Merge branch 'main' into chouquette/omnibus_cache

Pythyu committed:
feat(omnibus-cache): configure cache for remote updater OCI packages

Pythyu committed:
feat(omnibus-cache): configure cache for remote updater OCI packages

Pythyu committed:
feat(test): use omnibus basedir in cache

Pythyu committed:
feat(test): use two different omnibus cache variable

Pythyu committed:
feat(test): updater omnibus cache suffix

Pythyu committed:
feat(updater): added package version into cache path

Pythyu committed:
feat(updater): replace OMNIBUS_GIT_CACHE_SUFFIX with an agent.omnibus-build argument

Pythyu committed:
feat(agent.py): black python formatting

Pythyu committed:
feat(agent.py): lint-python use generator instead of map

chouquette commented on tasks/agent.py:
```diff
@@ -857,6 +858,132 @@ def _send_build_metrics(ctx, overall_duration):
+    environment variables to ensure we don't omit a variable
+    """
+
+    def env_filter(item):
```
Open question, should we move this out of the file? I fear it might change semi regularly especially at first, and it's quite a large chunk of code

Pythyu replied:
I agree, I'm moving the whole into a `lib/omnibus_cache.py` file

chouquette replied:
Thanks!

Pythyu committed:
feat(PR): apply suggestions

Pythyu committed:
feat(PR): apply suggestions

Pythyu committed:
Merge branch 'main' into chouquette/omnibus_cache

Pythyu committed:
feat(linter): fix flake8 unused python import

Pythyu committed:
feat(CODEOWNERS): update codeowners for agent-build-and-release to own omnibus_cache.py

spencergilbert commented on tasks/libs/omnibus_cache.py:
```diff
@@ -0,0 +1,127 @@
+def _get_build_images(ctx):
+    # We intentionally include both build images & their test suffixes in the pattern
+    # as a test image and the merged version shouldn't share their cache
+    tags = ctx.run("grep -E 'DATADOG_AGENT_.*BUILDIMAGES' .gitlab-ci.yml | cut -d ':' -f 2", hide='stdout').stdout
```
<!-- Suggestions are specific requests to improve the subject under review. It is assumed that we all want to do what's best, so these comments are never dismissed as “mere suggestions”, but are taken seriously.  -->
**💬 suggestion:** ‏I'd probably lean towards writing this in python directly, rather than shelling out. But, this isn't a blocking comment.

spencergilbert reviewed (commented):
Is there a link to a pipeline where this can be seen in action?

dustmop reviewed (approved):
LGTM for ASC owned files

KSerrania commented on tasks/agent.py:
```diff
@@ -933,6 +935,30 @@ def omnibus_build(
+    omnibus_cache_dir = os.environ.get('OMNIBUS_GIT_CACHE_DIR')
+    use_omnibus_git_cache = omnibus_cache_dir is not None
+    if use_omnibus_git_cache:
+        omnibus_cache_dir += install_directory
```
This seems fragile as a path concatenation operation, should that use [os.path.join](https://docs.python.org/3/library/os.path.html#os.path.join) instead?

KSerrania commented on tasks/agent.py:
```diff
@@ -933,6 +935,30 @@ def omnibus_build(
     with timed(quiet=True) as bundle_elapsed:
         bundle_install_omnibus(ctx, gem_path, env)
 
+    omnibus_cache_dir = os.environ.get('OMNIBUS_GIT_CACHE_DIR')
+    use_omnibus_git_cache = omnibus_cache_dir is not None
```
IIUC, `OMNIBUS_GIT_CACHE_DIR` is not set on the Windows Agent builds, meaning that the build cache is not enabled for Windows.

The PR description doesn't mention that this change is Linux-only, so could you confirm that this is expected? If yes, can this be added somewhere in code and in the PR description?

Pythyu committed:
feat(PR): apply suggestion, enable cache on windows

Pythyu committed:
fix(path): os path join fix

Pythyu commented on .gitlab/package_build/windows.yml:
```diff
@@ -21,6 +21,7 @@
       -e CI_JOB_NAME_SLUG=${CI_JOB_NAME_SLUG}
       -e CI_COMMIT_REF_NAME=${CI_COMMIT_REF_NAME}
       -e OMNIBUS_TARGET=${OMNIBUS_TARGET}
+      -e OMNIBUS_GIT_CACHE_DIR="/tmp/omnibus-git-cache"
```
not at all 🤦

Pythyu committed:
feat(windows): correct windows path

Pythyu committed:
Merge branch 'main' into chouquette/omnibus_cache

Pythyu committed:
fix(omnibus_cache): use aws.cmd on windows

Pythyu committed:
Merge branch 'chouquette/omnibus_cache' of github.com:DataDog/datadog-agent into chouquette/omnibus_cache

Pythyu committed:
fix(omnibus_cache): use correct git cache bundle path for windows

Pythyu committed:
fix(omnibus_cache): use correct git cache bundle path for windows

chouquette committed:
Merge branch 'main' into chouquette/omnibus_cache

chouquette committed:
fix conflict resolution mistake

chouquette committed:
omnibus cache: ignore ssh related env variables

chouquette committed:
correctly dispatch the install dir

chouquette committed:
remove unneeded path sanitization

os.path.join will handle this

chouquette committed:
debug

chouquette committed:
fixes & debug

chouquette committed:
don't provide an empty install path to the non-OCI installer builds

chouquette committed:
provide a mapping for the non-OCI installer

chouquette committed:
fix typo

chouquette committed:
reintroduce needed sanitization

chouquette committed:
add more patterns to environment exclusion list

chouquette committed:
add a comment explainig the cache dir & install path

Also, this will not cause the cache to be invalidated and can be used to
test if everything still works

chouquette committed:
filter more env variables

chouquette commented:
/merge

alopezz commented on tasks/omnibus.py:
```diff
@@ -219,6 +230,44 @@ def build(
     with timed(quiet=True) as bundle_elapsed:
         bundle_install_omnibus(ctx, gem_path, env)
 
+    omnibus_cache_dir = os.environ.get('OMNIBUS_GIT_CACHE_DIR')
+    use_omnibus_git_cache = omnibus_cache_dir is not None
+    if use_omnibus_git_cache:
```
I'm hoping we can implement this as separate functions or a context manager at some point.

chouquette replied:
That would make sense indeed

alopezz reviewed (approved):
Good work! Nothing blocking here.

KSerrania commented on tasks/libs/common/omnibus.py:
```diff
@@ -6,6 +7,151 @@
+    for k, v in environment.items():
+        print(f'\tUsing environment variable {k} to compute cache key')
+        h.update(str.encode(f'{k}={v}'))
+    # FIXME: include omnibus-ruby and omnibus-software version once they are pinned
```
I think this needs to be updated?

alopezz committed:
simplify install directory sanitization

chouquette commented on tasks/omnibus.py:
```diff
@@ -219,6 +230,44 @@ def build(
+        # /git/cache/dir/a/b/c/.git
+        if install_directory is None:
+            install_directory = install_dir_for_project(target_project)
+        if install_directory[0] == "/":
+            # Is the path starts with a /, it's considered the new root for the joined path
+            # which effectively drops whatever was in omnibus_cache_dir
+            install_directory = install_directory[1:]
```
Indeed, applied locally

chouquette committed:
remove old fixme

chouquette committed:
use omnibus commits in cache key

KSerrania commented on tasks/libs/common/omnibus.py:
```diff
@@ -1,9 +1,164 @@
+
+
+def _get_omnibus_commits(field):
+    release_version = os.environ['RELEASE_VERSION_7']
```
I think you want to check `RELEASE_VERSION`, and then  `RELEASE_VERSION_7` if the first one is not found, because of the way we set these variables currently.

Note for later: we should standardize all builds, make them set `RELEASE_VERSION` explicitly, to avoid having to use this hack.

chouquette replied:
Indeed this was causing a failure on windows.
I might have some questions about these variables during the summit, it's a bit unclear to me.

iglendd commented on tasks/winbuildscripts/dobuild.bat:
```diff
@@ -11,6 +11,7 @@ if NOT DEFINED GO_VERSION_CHECK set GO_VERSION_CHECK=%~4
 
 set OMNIBUS_BUILD=omnibus.build
 set OMNIBUS_ARGS=--python-runtimes "%PY_RUNTIMES%"
+set INSTALL_DIR=opt\datadog-agent
```
Did it work for windows build? Is it a new directory to be created ? In what parent directory?

chouquette replied:
It worked as this is only used for knowing the subdirectory in which to locate the cache, so it would be created in the value provided by `OMNIBUS_GIT_CACHE_DIR`, which points to `C:\TEMP\omnibus-git-cache` by default on Windows.
However, I believe this is actually not unneeded. I removed it and will check the pipeline results

chouquette committed:
handle both RELEASE_VERSION and RELEASE_VERSION_7

chouquette committed:
remove unneeded windows task parameter

KSerrania commented on tasks/libs/common/omnibus.py:
```diff
@@ -1,9 +1,167 @@
+    omnibus_last_commit = ctx.run('git log -n 1 --pretty=format:%H omnibus/', hide='stdout').stdout
+    h.update(str.encode(omnibus_last_commit))
+    print(f'\tLast omnibus commit is {omnibus_last_commit}')
+    buildimages_hash = _get_build_images(ctx)
+    for img_hash in buildimages_hash:
+        h.update(str.encode(img_hash))
```
Not needed for now, but for debugging purposes, you may want to log explicitly what buildimages entries go in the cache key, like you do for the environment variables.

KSerrania reviewed (approved):
LGTM. If you plan on merging this today, can you sync with @FlorentClarret? This will may conflict with the Python linter changes he's making in https://github.com/DataDog/datadog-agent/pull/24590, so he'll have to rebase.

chouquette committed:
display omnibus commits sha1

chouquette commented:
/merge

chouquette committed:
Merge branch 'main' into chouquette/omnibus_cache

chouquette commented:
/merge

chouquette commented:
/merge -c

chouquette commented:
/merge

iliakur reviewed (approved):
seems fine from agent ints perspective, tho would be nice to get rid of the repetitive `always_build true`
